app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,474 @@
+      end
+
+      # write the repository json file.
+      repository.write
+    end
+
+    # update the root json file to include/exlude this project.
+    update_root_file
+
+  end
+
+  def process_commit(ref)
+    if activated? && ref_exported?(ref)
+      package = ref_package(ref)
+      repository.add_package(package)
+    end
+  rescue Exception => e
+    # These errors are non-critical and have an impact on the exported
+    # packages. These errors can be ignored and are logged for troubleshooting.
+    log(e.message)
+  end
+
+  def update_root_file
+    root_json = Composer::Json::JsonFile.new(root_path)
+    root = root_json.read
+    key = "p/#{repo_filename}"
+    includes = root['includes'] || {}
+
+    if activated?
+      includes[key] ||= {}
+      includes[key]['sha1'] = Digest::SHA1.file(repo_path).hexdigest
+      includes = includes.sort_by{|k,v| k}.to_h
+    else
+      includes.delete(key)
+    end
+
+    root_json.write({ packages: [], includes: includes })
+  rescue Exception => e
+    # These errors are critical indicating that we can not process
+    # the root json file. Log and raise
+    log(e.message)
+    raise e
+  end
+
+  private
+
+  def ref_package(ref)
+    case package_mode
+    when 'default'
+      loader = Composer::Package::Loader::ProjectRootLoader.new
+      loader.load(project, ref)
+    when 'attributes'
+      loader = Composer::Package::Loader::ProjectAttributesLoader.new
+      loader.load(project, ref, package_type)
+    when 'advanced'
+      loader = Composer::Package::Loader::ProjectLoader.new
+      loader.load(project, ref, ActiveSupport::JSON.decode(custom_json))
+    end
+  end
+
+  def repository
+    @repository ||= Composer::Repository::ProjectRepository.new(
+      Composer::Json::JsonFile.new(repo_path)
+    )
+  end
+
+  def ref_exported?(ref)
+    if ref.instance_of?(Gitlab::Git::Branch)
+      branch_exported?(ref)
+    elsif ref.instance_of?(Gitlab::Git::Tag)
+      tag_exported?(ref)
+    else
+      false
+    end
+  end
+
+  def branch_exported?(branch)
+    if branch_filters
+      filters = branch_filters.strip! || branch_filters
+      filters = filters.gsub(' ', '').split(',')
+    else
+      filters = []
+    end
+    if filters.empty?
+      export_branches == '1'
+    else
+      export_branches == '1' && filters.include?(branch.name)
+    end
+  end
+
+  def tag_exported?(tag)
+    if tag_filters
+      filters = tag_filters.strip! || tag_filters
+      filters = filters.gsub(' ', '').split(',')
+    else
+      filters = []
+    end
+    if filters.empty?
+      export_tags == '1'
+    else
+      export_tags == '1' && filters.include?(tag.name)
+    end
+  end
+
+  def output_dir
+    Rails.public_path
+  end
+
+  def provider_dir
+    File.join(output_dir, '/p');
+  end
+
+  def root_filename
+    'packages.json'
+  end
+
+  def root_path
+    File.join(output_dir, root_filename)
+  end
+
+  def repo_filename
+    "project-#{project.id}.json"
```

Comment:
Backup rake task is committed. The events for updating the project attributes are pending.